### ISSUE TITLE
StepVerifier hasOperatorErrorMatching failing to count operators properly

### ISSUE THREAD

**Issue by fjsventura** [reactor/reactor-core#2991]
After upgrading from reactor 3.1.10.RELEASE (Spring-Boot 2.0.6.RELEASE) to reactor 3.4.14 (Spring-Boot 2.6.3) the StepVerifier started failing in some error assertions with `.hasOperatorErrorMatching()`. Particularly, streams that had only one Mono.error being created and only a single operation to handle it failed with "Expected exactly one operator error, 2 found." and similars.

## Expected Behavior
If there is a single "onError" step in a stream, the `.hasOperatorErrorMatching()` should be called exactly only once.

## Actual Behavior
Depending on the stream the `.hasOperatorErrorMatching()` method may be called more than once and fail the assertion, even if there is a single "onError" step in a stream.

## Steps to Reproduce
To reproduce the following snippet should be enough:

```java
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

class Scratch {

    public static void main(String[] args) {
        Flux flux = Flux
            .just(">>>>>ABC<<<<<")
            .flatMap(s -> Mono.just(">>>zyx<<<"))
            .flatMap(o -> Mono.error(new RuntimeException("WOW!")))
            .onErrorResume(throwable -> Mono.empty())
            .log();

        StepVerifier.create(flux)
            .expectNextCount(0)
            .expectComplete()
            .verifyThenAssertThat()
            .hasOperatorErrorMatching(RuntimeException.class::isInstance).hasOperatorErrorWithMessage("WOW!");
    }
}
```

In this case, if we comment the `.flatMap(s -> Mono.just(">>>zyx<<<"))`  the StepVerifier assertion passes and if not commented it fails, even though there is a single "onError" step and only a single Mono.error being created. 

## Your Environment
The relevant parts of my pom:
```xml
	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>2.6.3</version>
	</parent>
	<properties>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
		<project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
		<java.version>1.8</java.version>
	</properties>
<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-webflux</artifactId>
		</dependency>
		<dependency>
			<groupId>org.projectlombok</groupId>
			<artifactId>lombok</artifactId>
		</dependency>

		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>
		<dependency>
			<groupId>io.projectreactor</groupId>
			<artifactId>reactor-test</artifactId>
			<scope>test</scope>
		</dependency>
		<dependency>
			<groupId>io.projectreactor</groupId>
			<artifactId>reactor-test</artifactId>
		</dependency>
	</dependencies>
```

* Reactor version(s) used: reactor-core and reactor-test 3.4.14 (From Spring-Boot-Webflux-Starter 3.6.3)
* Other relevant libraries versions (eg. `netty`, ...): All the dependencies are above
* JVM version (`java -version`): openjdk version "1.8.0_312"
* OS and version (eg `uname -a`): Linux PT-PF2J01WF 5.13.0-35-generic

**Comment by simonbasle** [comment]
Judging from the title of this issue ("failing to count operators properly") I'm not sure the `hasOperatorErrorMatching` does what you think it does.

 1. there is no relationship between the number of `Mono.error` or `onErrorXxx` operator usage and this assertion
 2. assertion methods post `.verifyThenAssertThat()` are more about internals and hooks, which you generally care less about as a user

The `Assertions#hasOperatorErrorXxx` assertions evaluate the number of times the `Hooks.onOperatorError` hook has been invoked. This hook is about globally transforming exceptions, eg. to wrap some specific business exception into a different type of exception. Generally this type of hook should be idempotent.

Since `3.2.0.RELEASE`, support for `onErrorContinue` has been added. When an error occurs, operators that support onErrorContinue will pass the exception to a dedicated hook which can suppress the exception by returning `null`. If no such hook is defined, it actually defaults to applying the more generic `onOperatorError` hook.
In most operators, in addition to handling the onErrorContinue case, an exception will always go through the `onOperatorError` hook.

This is what happens here. No `onErrorContinue` is set up, so the exception is passed to the `onOperatorError` hook twice: once as the default `onErrorContinue` hook and once as the standard "there's been an error" hook.

This has never been an issue since 2018, likely because most `onOperatorError` hooks out there are idempotent.

TL;DR: either accommodate the fact that hooks can be invoked twice, or rework the test to avoid using `verifyThenAssertThat()` altogether

I'd suggest:
```java
Flux flux = Flux
	.just(">>>>>ABC<<<<<")
	.flatMap(s -> Mono.just(">>>zyx<<<"))
	.flatMap(o -> Mono.error(new RuntimeException("WOW!")))
	.onErrorResume(throwable -> Mono.just("EXCEPTION: " + throwable.getMessage()))
	.log();

StepVerifier.create(flux)
	.expectNext("Exception: WOW!")
	.expectComplete()
	.verify();
```

**Comment by fjsventura** [comment]
Our expectation was that `.hasOperatorErrorMatching()` would be called once for each "onError" step. For instance in this case, the `.onErrorResume()` would be one call and if we added something like `.onErrorMap()` it would be a second call. We already did exactly what you suggested. Thank you very much
